src/Nethermind/Nethermind.TxPool/TxPool.cs

```diff
@@ -338,6 +310,171 @@ private AddTxResult AddCore(Transaction tx, bool isPersistentBroadcast, bool isR
+                    isKnown |= !inserted;
+                }
+            }
+
+            if (!isKnown)
+            {
+                isKnown |= !isReorg && (_txStorage.Get(tx.Hash) is not null);
+            }
+
+            if (isKnown)
+            {
+                // If transaction is a bit older and already known then it may be stored in the persistent storage.
+                Metrics.PendingTransactionsKnown++;
+                if (_logger.IsTrace) _logger.Trace($"Skipped adding transaction {tx.ToString("  ")}, already known.");
+                return AddTxResult.AlreadyKnown;
+            }
+            
+            HandleOwnTransaction(tx, isPersistentBroadcast);
+            NotifySelectedPeers(tx);
+            StoreTx(tx);
+            NewPending?.Invoke(this, new TxEventArgs(tx));
+            return AddTxResult.Added;
+        }
+
+        private void UpdateBucket(Address senderAddress)
+        {
+            var bucketSnapshot = _transactions.GetBucketSnapshot(senderAddress);
+            
+            if (bucketSnapshot.Length == 0)
+            {
+                return;
+            }
+            
+            Account? account = _stateProvider.GetAccount(senderAddress);
+            UInt256 balance = account?.Balance ?? UInt256.Zero;
+            long currentNonce = (long)(account?.Nonce ?? UInt256.Zero);
+
+            UpdateGasBottleneck(bucketSnapshot, currentNonce, balance);
+        }
+
+        private void UpdateGasBottleneck(IReadOnlyList<Transaction> bucketSnapshot, long currentNonce, UInt256 balance)
+        {
+            Transaction tx = bucketSnapshot[0];
+            UInt256 previousTxBottleneck = tx.IsEip1559 ? CalculatePayableGasPrice(tx, balance) : tx.GasPrice;
+
+            for (int i = 0; i < bucketSnapshot.Count; i++)
+            {
+                tx = bucketSnapshot[i];
+                UInt256 gasBottleneck = 0;
+
+                if (tx.Nonce == currentNonce + i)
+                {
+                    UInt256 effectiveGasPrice = tx.IsEip1559 ? UInt256.Min(tx.FeeCap, tx.GasPremium + CurrentBaseFee) : tx.GasPrice;
+                    gasBottleneck = UInt256.Min(effectiveGasPrice, previousTxBottleneck);
+                }
+                
+                _transactions.NotifyChange(tx.Hash, tx, t =>
+                {
+                    tx.GasBottleneck = gasBottleneck;
+                });
+                
+                previousTxBottleneck = gasBottleneck;
+            }
+        }
+
+        private UInt256 CalculatePayableGasPrice(Transaction tx, UInt256 balance)
+        {
+            UInt256 payableGasPrice;
+            
+            if (balance > tx.Value)
+            {
+                UInt256 effectiveGasPrice = UInt256.Min(tx.FeeCap, tx.GasPremium + CurrentBaseFee);
+                UInt256 balanceAvailableForFeePayment = balance - tx.Value;
+                balanceAvailableForFeePayment.Divide((UInt256)tx.GasLimit, out UInt256 maxPayablePricePerGasUnit);
```

Comment:
I'm just wondering is there any risk for division by zero?

Reply:
I will add one more check to be sure there will be no surprises